server/engine/src/main/java/org/eclipse/lsp/cobol/core/engine/processor/AstProcessor.java

```diff
@@ -74,20 +77,25 @@ public void process(ProcessingPhase phase, Node node, ProcessingContext ctx) {
   /**
    * Process tree node and its children after tree construction.
    *
-   * @param processor list of available processors
+   * @param processors list of available processors
    * @param node a node to process
    * @param ctx processing context
    */
   private void process(Map<Class<? extends Node>, List<Processor<? extends Node>>> processors,
       Node node, ProcessingContext ctx) {
     ThreadInterruptionUtil.checkThreadInterrupted();
     final Class<? extends Node> nodeClass = node.getClass();
-    processors.forEach((key, value) -> {
-      if (!key.isAssignableFrom(nodeClass))
-        return;
-      for (Processor<? extends Node> p : value)
-        ((Processor<Node>) p).accept(node, ctx);
-    });
-    node.getChildren().forEach(n -> process(processors, n, ctx));
+    if (nodeClass == ProgramNode.class) {
+      ctx.getCurrentProgramNodeStack().push((ProgramNode) node);
+    }
+    for (Processor<? extends Node> processor : processors.getOrDefault(nodeClass, Collections.emptyList())) {
```

Comment:
Doesn't this fail for the cases where nodeClass used in registration is a parent class of the node?

Reply:
The new version looks more error prone to me. Should we require the class being registered being final and maybe create a registerBase method for the case where base is not abstract?

Reply:
Abstract check added in https://github.com/eclipse-che4z/che-che4z-lsp-for-cobol/pull/2663/commits/4f81424e32c941b0cb5fb26ccfe95c5735a7e9ea